File: surf-api-core/surf-api-core-server/src/main/java/dev/slne/surf/surfapi/core/server/listener/packet/chat/ChatPacketListener.java
package dev.slne.surf.surfapi.core.server.listener.packet.chat;

import com.github.retrooper.packetevents.event.PacketListenerAbstract;
import com.github.retrooper.packetevents.event.PacketListenerPriority;
import com.github.retrooper.packetevents.event.PacketSendEvent;
import com.github.retrooper.packetevents.protocol.packettype.PacketType.Play.Server;
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;

public final class ChatPacketListener extends PacketListenerAbstract {

  private static final PlainTextComponentSerializer serializer = PlainTextComponentSerializer.plainText();

  public ChatPacketListener(PacketListenerPriority priority) {
    super(priority);
  }

  @Override
  public void onPacketSend(
      PacketSendEvent event) { // TODO: 02.03.2024 00:34 - packet events need to fix their wrapper
    if (event.getPacketType() != Server.SYSTEM_CHAT_MESSAGE) {
      return;
    }

//    final WrapperPlayServerSystemChatMessage packet = new WrapperPlayServerSystemChatMessage(event);
//    final String message = serializer.serialize(packet.getMessage());

//    System.out.println("Message: " + message);
//
//    if (!message.startsWith(ComponentMessage.COMPONENT_MESSAGE_PREFIX)) {
//      return;
//    }
//
//    System.out.println("Message starts with prefix");
//
//    final Component coloredMessage = GsonComponentSerializer.gson()
//        .deserialize(message.substring(ComponentMessage.COMPONENT_MESSAGE_PREFIX.length()));
//
//    packet.setMessage(coloredMessage);
  }
}

File: surf-api-core/surf-api-core-server/src/main/kotlin/dev/slne/surf/surfapi/core/server/listener/packet/chat/ChatPacketListener.kt
package dev.slne.surf.surfapi.core.server.listener.packet.chat

import com.github.retrooper.packetevents.event.PacketListenerAbstract
import com.github.retrooper.packetevents.event.PacketListenerPriority
import com.github.retrooper.packetevents.event.PacketSendEvent
import com.github.retrooper.packetevents.protocol.packettype.PacketType.Play
import com.github.retrooper.packetevents.wrapper.play.server.WrapperPlayServerSystemChatMessage
import dev.slne.surf.surfapi.core.api.messages.ComponentMessage
import dev.slne.surf.surfapi.core.api.messages.adventure.plain
import net.kyori.adventure.text.serializer.gson.GsonComponentSerializer

class ChatPacketListener(priority: PacketListenerPriority) : PacketListenerAbstract(priority) {
    override fun onPacketSend(event: PacketSendEvent) {
        if (event.packetType != Play.Server.SYSTEM_CHAT_MESSAGE) {
            return
        }

        val packet = WrapperPlayServerSystemChatMessage(event)
        val message = packet.message.plain()

        if (!message.startsWith(ComponentMessage.COMPONENT_MESSAGE_PREFIX)) {
            return
        }

        val coloredMessage = GsonComponentSerializer.gson()
            .deserialize(message.substring(ComponentMessage.COMPONENT_MESSAGE_PREFIX.length))

        packet.message = coloredMessage
        event.markForReEncode(true)
    }
}
